### PR TITLE
Generate S/Kademlia compatible keypairs and node IDs for discovery plugin.

### PR DESCRIPTION
**Pull request by jack0**

### CODE REVIEW

File: examples/basic/example_test.go [hunk 1]
```diff
@@ -47,10 +47,16 @@ func ExampleBasicPlugin() {
...
 		builder := network.NewBuilder()
-		builder.SetKeys(ed25519.RandomKeyPair())
-		builder.SetAddress(network.FormatAddress("tcp", host, uint16(startPort+i)))
-
-		builder.AddPlugin(new(discovery.Plugin))
+		address := network.FormatAddress("tcp", host, uint16(startPort+i))
+		kp, id := discovery.GenerateKeyPairAndID(address)
+		builder.SetKeys(kp)
+		builder.SetAddress(address)
+		builder.SetNodeIDNonce(peer.GetNonce(id))
```
Comment (iwasaki-kenta) [inline examples/basic/example_test.go]: Hmm, we shouldn't keep S/Kademlia-related methods in network.Builder; it's better we have discovery.GenerateKeypairAndID(address) already generate the nonce.

File: network/discovery/identity.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	"github.com/perlin-network/noise/crypto/blake2b"
+	"github.com/perlin-network/noise/crypto/ed25519"
+	"github.com/perlin-network/noise/peer"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	// number of preceding bits of 0 in the H(H(key_public)) for NodeID generation
+	c1 = 16
```
Comment (iwasaki-kenta) [inline network/discovery/identity.go]: We should expose C1 and C2 to be configurable.

File: examples/basic/example_test.go (hunk 1 above)
Comment (jack0) [reply examples/basic/example_test.go]: the builder is the one that creates the Peer though so it needs to know at least whether the peer needs to have a s/kademlia compatible node id

File: network/network.go [hunk 3]
```diff
@@ -457,9 +458,11 @@ func (n *Network) Accept(incoming net.Conn) {
 			}
 
 			client.ID = (*peer.ID)(msg.Sender)
+			// resolves hostnames
+			clientAddr, err = ToUnifiedAddress(client.ID.Address)
```
Comment (doug-perlin) [inline network/network.go]: Do you have to check the error before using clientAddr?

File: examples/basic/example_test.go [hunk 4]
```diff
@@ -47,11 +47,20 @@ func ExampleBasicPlugin() {
...
 
-		builder.AddPlugin(new(discovery.Plugin))
+		// enforce S/Kademlia node IDs
+		plugin := discovery.New(
+			discovery.WithPuzzleEnabled(discovery.DefaultC1, discovery.DefaultC2),
+		)
+		kp, nonce := plugin.PerformPuzzle()
 
+		builder.SetKeys(kp)
+		id := peer.CreateID(address, kp.PublicKey)
```
Comment (iwasaki-kenta) [inline examples/basic/example_test.go]: Care to modify peer.CreateID to support options s.t. `peer.CreateID(address, kp.PublicKey, peer.WithNonce(nonce), ...otherOptions)`?

**Pull request closed without merging**